File: src/api/fileReview.ts
import { invoke } from "@tauri-apps/api/core";

export interface FileReviewFile {
  path: string;
  action: string;
  moveTo?: string;
  baseContent?: string;
  candidateContent?: string;
  editedContent?: string;
  keep: boolean;
}

export interface FileReviewSession {
  threadId: string;
  callId: string;
  rawPatch: string;
  createdAtMs: number;
  updatedAtMs: number;
  files: FileReviewFile[];
}

export interface FileReviewActionResponse {
  ok: boolean;
  message: string;
  changedFiles: Array<{ path: string; action: string }>;
}

/**
 * 获取后端缓存的待审阅会话。
 *
 * 注意：该数据来自 apply_patch 的“写盘前”阶段，文件尚未真正落盘。
 */
export async function fileReviewGet(
  threadId: string,
  callId: string,
): Promise<FileReviewSession> {
  return invoke("file_review_get", { threadId, callId });
}

/**
 * 更新单文件审阅数据（keep 勾选或编辑后的内容）。
 *
 * 约束：
 * - 对 deleted 文件传 editedContent 会返回后端校验错误；
 * - editedContent 太长会返回后端长度限制错误。
 */
export async function fileReviewUpdate(
  threadId: string,
  callId: string,
  path: string,
  keep?: boolean,
  editedContent?: string,
): Promise<FileReviewSession> {
  return invoke("file_review_update", { threadId, callId, path, keep, editedContent });
}

/**
 * 应用审阅结果并真正写盘。
 *
 * keepAll 为 true 时会忽略 keepPaths，直接应用全部候选文件。
 */
export async function fileReviewApply(
  threadId: string,
  callId: string,
  keepAll?: boolean,
  keepPaths?: string[],
): Promise<FileReviewActionResponse> {
  return invoke("file_review_apply", { threadId, callId, keepAll, keepPaths });
}

/**
 * 取消审阅并丢弃后端缓存，不写入任何文件。
 */
export async function fileReviewCancel(
  threadId: string,
  callId: string,
): Promise<FileReviewActionResponse> {
  return invoke("file_review_cancel", { threadId, callId });
}

File: src-tauri/src/commands/file_review.rs
use std::path::PathBuf;

use serde::Serialize;
use tauri::{AppHandle, Emitter, State};

use crate::error::{AppError, AppResult};
use crate::file_review::{
    PendingPatchReview, ReviewApplyRequest, apply_review_to_workspace, pending_review_key,
    update_review_file,
};
use crate::state::AppState;
use crate::thread_store::FileChange;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FileReviewActionResponse {
    pub ok: bool,
    pub message: String,
    #[serde(default)]
    pub changed_files: Vec<FileChange>,
}

fn emit_review_updated(app: &AppHandle, payload: serde_json::Value) {
    app.emit("file-review-updated", payload.clone()).ok();
    crate::mobile_server::broadcast("file-review-updated", payload);
}

#[tauri::command]
pub async fn file_review_get(
    state: State<'_, AppState>,
    thread_id: String,
    call_id: String,
) -> AppResult<PendingPatchReview> {
    let key = pending_review_key(&thread_id, &call_id);
    let sessions = state.file_review_sessions.read().await;
    let Some(review) = sessions.get(&key) else {
        return Err(AppError::Custom(format!(
            "pending review not found for thread={} call={}",
            thread_id, call_id
        )));
    };
    Ok(review.clone())
}

#[tauri::command]
pub async fn file_review_update(
    app_handle: AppHandle,
    state: State<'_, AppState>,
    thread_id: String,
    call_id: String,
    path: String,
    keep: Option<bool>,
    edited_content: Option<String>,
) -> AppResult<PendingPatchReview> {
    let key = pending_review_key(&thread_id, &call_id);
    let updated = {
        let mut sessions = state.file_review_sessions.write().await;
        let Some(review) = sessions.get_mut(&key) else {
            return Err(AppError::Custom(format!(
                "pending review not found for thread={} call={}",
                thread_id, call_id
            )));
        };
        update_review_file(review, &path, keep, edited_content).map_err(AppError::Custom)?;
        review.clone()
    };

    emit_review_updated(
        &app_handle,
        serde_json::json!({
            "threadId": thread_id,
            "callId": call_id,
            "status": "updated",
        }),
    );
    Ok(updated)
}

#[tauri::command]
pub async fn file_review_apply(
    app_handle: AppHandle,
    state: State<'_, AppState>,
    thread_id: String,
    call_id: String,
    keep_all: Option<bool>,
    keep_paths: Option<Vec<String>>,
) -> AppResult<FileReviewActionResponse> {
    let key = pending_review_key(&thread_id, &call_id);
    let review = {
        let sessions = state.file_review_sessions.read().await;
        let Some(review) = sessions.get(&key) else {
            return Err(AppError::Custom(format!(
                "pending review not found for thread={} call={}",
                thread_id, call_id
            )));
        };
        review.clone()
    };

    // 使用当前工作目录作为 apply 目标根目录，保证命令层与工具执行层行为一致。
    let cwd = PathBuf::from(state.cwd.read().await.clone());
    let apply_result = apply_review_to_workspace(
        &cwd,
        &review,
        &ReviewApplyRequest {
            keep_all: keep_all.unwrap_or(false),
            keep_paths: keep_paths.unwrap_or_default(),
        },
    )
    .map_err(AppError::Custom)?;

    {
        let mut sessions = state.file_review_sessions.write().await;
        sessions.remove(&key);
    }

    emit_review_updated(
        &app_handle,
        serde_json::json!({
            "threadId": thread_id,
            "callId": call_id,
            "status": "applied",
            "changedFiles": apply_result.changed_files,
        }),
    );

    Ok(FileReviewActionResponse {
        ok: true,
        message: apply_result.message,
        changed_files: apply_result.changed_files,
    })
}

#[tauri::command]
pub async fn file_review_cancel(
    app_handle: AppHandle,
    state: State<'_, AppState>,
    thread_id: String,
    call_id: String,
) -> AppResult<FileReviewActionResponse> {
    let key = pending_review_key(&thread_id, &call_id);
    let removed = {
        let mut sessions = state.file_review_sessions.write().await;
        sessions.remove(&key)
    };
    if removed.is_none() {
        return Err(AppError::Custom(format!(
            "pending review not found for thread={} call={}",
            thread_id, call_id
        )));
    }

    emit_review_updated(
        &app_handle,
        serde_json::json!({
            "threadId": thread_id,
            "callId": call_id,
            "status": "cancelled",
        }),
    );

    Ok(FileReviewActionResponse {
        ok: true,
        message: "Review cancelled.".to_string(),
        changed_files: Vec::new(),
    })
}
